torchnet/meter/apmeter.py

```diff
@@ -0,0 +1,135 @@
+import math
+from . import meter
+import numpy as np
+import torch
+
+
+class APMeter(meter.Meter):
+    """
+    The APMeter measures the average precision per class.
+
+    The APMeter is designed to operate on `NxK` Tensors `output` and
+    `target`, and optionally a `Nx1` Tensor weight where (1) the `output`
+    contains model output scores for `N` examples and `K` classes that ought to
+    be higher when the model is more convinced that the example should be
+    positively labeled, and smaller when the model believes the example should
+    be negatively labeled (for instance, the output of a sigmoid function); (2)
+    the `target` contains only values 0 (for negative examples) and 1
+    (for positive examples); and (3) the `weight` ( > 0) represents weight for
+    each sample.
+    """
+    def __init__(self):
+        self.reset()
+
+    def reset(self):
+        """Resets the meter with empty member variables"""
+        self.scores = torch.FloatTensor(torch.FloatStorage())
+        self.targets = torch.LongTensor(torch.LongStorage())
+        self.weights = torch.FloatTensor(torch.FloatStorage())
+
+    def add(self, output, target, weight=None):
+        """
+        Args:
+            output (Tensor): NxK tensor that for each of the N examples
+                indicates the probability of the example belonging to each of
+                the K classes, according to the model. The probabilities should
+                sum to one over all classes
+            target (Tensor): binary NxK tensort that encodes which of the K
+                classes are associated with the N-th input
+                    (eg: a row [0, 1, 0, 1] indicates that the example is
+                         associated with classes 2 and 4)
+            weight (optional, Tensor): Nx1 tensor representing the weight for
+                each example (each weight > 0)
+        """
+        if not torch.is_tensor(output):
+            output = torch.from_numpy(output)
+        if not torch.is_tensor(target):
+            target = torch.from_numpy(target)
+
+        if weight is not None:
+            if not torch.is_tensor(weight):
+                weight = torch.from_numpy(weight)
+            weight = weight.squeeze()
+        if output.dim() == 1:
+            output = output.view(output.size(0), 1)
```

Comment:
either `output.view(-1, 1)` or `output.unsqueeze(-1)` would be more readable here